### PR TITLE
[CMake][NFC] Introduce component targets for proper dependency tracking

### PR DESCRIPTION
**Pull request by bulbazord**
This commit introduces a CMake target for each component, adds install targets
for them, and switches build-script-impl to use the target `install-components`
for installation.
This is equivalent to what already exists, because right now install rules are
only generated for components that we want to install. Therefore, this commit
should be an NFC.

### CODE REVIEW

**Comment by xiaobai** [comment]
cc @compnerd @gottesmm @Rostepher

**Comment by compnerd** [comment]
@swift-ci please smoke test

**Comment by compnerd** [comment]
@swift-ci please smoke test

**Comment by xiaobai** [comment]
Friendly ping! @compnerd @gottesmm @Rostepher

**Comment by xiaobai** [comment]
ping @compnerd @gottesmm @Rostepher

**Comment by compnerd** [comment]
@swift-ci please clean test

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
Seems like we may be missing the `module.modulemap` for glibc?

**Comment by xiaobai** [comment]
> Seems like we may be missing the `module.modulemap` for glibc?

I just built and installed with this patch and without this patch. Both have a `glibc.modulemap` in the exact same place, so I don't think that's the problem. I'm running the test suite on my Linux box now, will see if I can reproduce the failure easily.

**Comment by xiaobai** [comment]
@swift-ci please smoke test

**Comment by xiaobai** [comment]
@swift-ci please smoke test

**Comment by xiaobai** [comment]
@swift-ci please smoke test

**Comment by xiaobai** [comment]
@swift-ci please smoke test Linux platform

**Comment by xiaobai** [comment]
@swift-ci please smoke test

**Comment by xiaobai** [comment]
@swift-ci please smoke test

**Comment by xiaobai** [comment]
@swift-ci please test

**Comment by xiaobai** [comment]
@swift-ci please smoke test

**Comment by xiaobai** [comment]
@swift-ci please smoke test

**Comment by xiaobai** [comment]
@swift-ci please smoke test

**Comment by xiaobai** [comment]
@swift-ci please smoke test Linux platform

**Comment by xiaobai** [comment]
@swift-ci please smoke test

**Comment by xiaobai** [comment]
friendly ping! @compnerd @Rostepher @gottesmm

File: CMakeLists.txt [hunk 1]
```diff
@@ -1051,6 +1051,8 @@ if(SWIFT_NEED_EXPLICIT_LIBDISPATCH)
   else()
     set(SOURCEKIT_RUNTIME_DIR lib)
   endif()
+  add_dependencies(sourcekit-inproc dispatch)
+  add_dependencies(sourcekit-inproc BlocksRuntime)
```
Comment (compnerd) [inline CMakeLists.txt]: Might be nice to have this be a single call.

File: CMakeLists.txt [hunk 2]
```diff
@@ -499,6 +499,16 @@ include(CMakePushCheckState)
...
 
+include(SwiftSharedCMakeConfig)
+
+# Support building Swift as a standalone project, using LLVM as an
+# external library.
+if(SWIFT_BUILT_STANDALONE)
+  swift_common_standalone_build_config(SWIFT)
+else()
+  swift_common_unified_build_config(SWIFT)
+endif()
```
Comment (compnerd) [inline CMakeLists.txt]: Might be nice to have a comment to explain that this needs to occur early to get the LLVM CMake macros (so someone doesn't try to move this around).

File: cmake/modules/AddSwift.cmake [hunk 3]
```diff
@@ -2437,3 +2445,9 @@ macro(add_swift_tool_symlink name dest component)
   add_llvm_tool_symlink(${name} ${dest} ALWAYS_GENERATE)
   llvm_install_symlink(${name} ${dest} ALWAYS_GENERATE COMPONENT ${component})
 endmacro()
+
+function(add_swift_install_targets target component)
```
Comment (compnerd) [inline cmake/modules/AddSwift.cmake]: Is there a reason for creating this really?  It seems that we could just directly use `add_llvm_install_targets`.

File: cmake/modules/SwiftComponents.cmake [hunk 4]
```diff
@@ -92,11 +92,21 @@ macro(swift_configure_components)
...
+    add_swift_install_targets(${component} ${component})
+
     string(TOUPPER "${component}" var_name_piece)
     string(REPLACE "-" "_" var_name_piece "${var_name_piece}")
     set(SWIFT_INSTALL_${var_name_piece} FALSE)
   endforeach()
 
+  # NOTE: never-install is a dummy component to indicate something should not
+  # be installed. We explicitly do not add an install target for this.
+  add_custom_target(never_install)
```
Comment (compnerd) [inline cmake/modules/SwiftComponents.cmake]: Bleh, I think that we should remove the `never_install` target instead.

File: cmake/modules/SwiftComponents.cmake [hunk 5]
```diff
@@ -92,11 +92,21 @@ macro(swift_configure_components)
...
+
+  add_custom_target(components)
```
Comment (compnerd) [inline cmake/modules/SwiftComponents.cmake]: I think that `swift-components` is better here.  Consider cases of unified builds, `components` does not identify the subproject which is providing the components.

File: tools/SourceKit/tools/complete-test/CMakeLists.txt [hunk 6]
```diff
@@ -24,6 +24,7 @@ if(SWIFT_ANALYZE_CODE_COVERAGE)
     LINK_FLAGS " -fprofile-instr-generate -fcoverage-mapping")
 endif()
 
+add_dependencies(tools complete-test)
```
Comment (compnerd) [inline tools/SourceKit/tools/complete-test/CMakeLists.txt]: Why are testing tools being installed? :/

File: tools/SourceKit/tools/sourcekitd-test/CMakeLists.txt [hunk 7]
```diff
@@ -32,6 +32,7 @@ if(SWIFT_ANALYZE_CODE_COVERAGE)
     LINK_FLAGS " -fprofile-instr-generate -fcoverage-mapping")
 endif()
 
+add_dependencies(tools sourcekitd-test)
```
Comment (compnerd) [inline tools/SourceKit/tools/sourcekitd-test/CMakeLists.txt]: And another instance ...

**Review by compnerd — APPROVED** [review]
I think that the cleanup of the testing tools for SourceKit is beyond the scope of this PR.  It would be nice to have the additional comments and some of the naming things fixed up.  You can do that here or in a follow up.

File: cmake/modules/SwiftComponents.cmake (hunk 4 above)
Comment (xiaobai) [reply cmake/modules/SwiftComponents.cmake]: I had a patch to make `INSTALL_IN_COMPONENT` optional, but I don't think @Rostepher liked it too much. 😛

**Comment by xiaobai** [comment]
@swift-ci please smoke test and merge

**Comment by xiaobai** [comment]
@swift-ci please smoke test and merge

**Comment by benlangmuir** [comment]
We've seen failures after this PR that look like we're no longer installing the clang builtin header symlink correctly.

https://ci.swift.org/job/oss-swift-incremental-RA-osx/8447/

```
indexstore-db: error: manifest parse error(s):
/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/module.modulemap:118:11: error: header 'stdarg.h' not found
   header "stdarg.h" // note: supplied by the compiler
          ^
```

I'm inclined to revert this PR.

**Comment by benlangmuir** [comment]
Comparing two recent PR tests:
https://ci.swift.org/job/swift-PR-osx-smoke-test/16435/ (passed)
https://ci.swift.org/job/swift-PR-osx-smoke-test/16437/ (failed)

It looks like after this PR merged, any new clean builds failed.  That's consistent with us not correctly installing the clang builtin header symlink, since an incremental build might have still had it sitting around.

**Comment by xiaobai** [comment]
> Comparing two recent PR tests:
> https://ci.swift.org/job/swift-PR-osx-smoke-test/16435/ (passed)
> https://ci.swift.org/job/swift-PR-osx-smoke-test/16437/ (failed)
> 
> It looks like after this PR merged, any new clean builds failed. That's consistent with us not correctly installing the clang builtin header symlink, since an incremental build might have still had it sitting around.

I see. I'll dig into it now and see what I can do. I'm invested in having this patch in, so I'll see what I can do. Revert if you absolutely must, I'll just resubmit after I fix.

**Comment by benlangmuir** [comment]
Reverted https://github.com/apple/swift/pull/26574 since this was starting to break lots of incremental and PR jobs.  When you re-land this, I suggest testing:
* smoke test
* clean test
* build toolchain

Thanks!

**Pull request merged**